// @onflow/config/dist/config.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@onflow/util-actor'), require('@onflow/util-logger'), require('@onflow/util-invariant')) :
  typeof define === 'function' && define.amd ? define(['exports', '@onflow/util-actor', '@onflow/util-logger', '@onflow/util-invariant'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.onflowConfig = {}, global.utilActor, global.logger, global.utilInvariant));
})(this, (function (exports, utilActor, logger, utilInvariant) { 'use strict';

  function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
      Object.keys(e).forEach(function (k) {
        if (k !== 'default') {
          var d = Object.getOwnPropertyDescriptor(e, k);
          Object.defineProperty(n, k, d.get ? d : {
            enumerable: true,
            get: function () { return e[k]; }
          });
        }
      });
    }
    n.default = e;
    return Object.freeze(n);
  }

  var logger__namespace = /*#__PURE__*/_interopNamespaceDefault(logger);

  /**
   * Fork network configuration
   * Allows a custom network to inherit contract aliases from a parent network
   */

  /**
   * Network configuration - either a string (host) or fork configuration
   */

  const pipe = function () {
    for (var _len = arguments.length, funcs = new Array(_len), _key = 0; _key < _len; _key++) {
      funcs[_key] = arguments[_key];
    }
    return v => {
      return funcs.reduce((res, func) => {
        return func(res);
      }, v);
    };
  };

  /***
   * Merge multiple functions returning objects into one object.
   * @param funcs - Functions to merge
   * @return Merged object
   */
  const mergePipe = function () {
    for (var _len2 = arguments.length, funcs = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      funcs[_key2] = arguments[_key2];
    }
    return v => {
      return funcs.reduce((res, func) => {
        return {
          ...res,
          ...func(v)
        };
      }, {});
    };
  };

  /**
   * @description Object check
   * @param value - Value to check
   * @returns Is object status
   */
  const isObject = value => value && typeof value === "object" && !Array.isArray(value);

  /**
   * @description Deep merge multiple objects.
   * @param target - Target object
   * @param sources - Source objects
   * @returns Merged object
   */
  const mergeDeep = function (target) {
    for (var _len3 = arguments.length, sources = new Array(_len3 > 1 ? _len3 - 1 : 0), _key3 = 1; _key3 < _len3; _key3++) {
      sources[_key3 - 1] = arguments[_key3];
    }
    if (!sources.length) return target;
    const source = sources.shift();
    if (isObject(target) && isObject(source)) {
      for (const key in source) {
        if (isObject(source[key])) {
          if (!target[key]) Object.assign(target, {
            [key]: {}
          });
          mergeDeep(target[key], source[key]);
        } else {
          Object.assign(target, {
            [key]: source[key]
          });
        }
      }
    }
    return mergeDeep(target, ...sources);
  };

  /**
   * @description Deep merge multiple Flow JSON.
   * @param value - Flow JSON or array of Flow JSONs
   * @returns Merged Flow JSON
   */
  const mergeFlowJSONs = value => Array.isArray(value) ? mergeDeep({}, ...value) : value;

  /**
   * @description Filter out contracts section of flow.json.
   * @param obj - Flow JSON
   * @returns Contracts section of Flow JSON
   */
  const filterContracts = obj => obj.contracts ? obj.contracts : {};

  /**
   * @description Filter out dependencies section of flow.json.
   * @param obj - Flow JSON
   * @returns Dependencies section of Flow JSON
   */
  const filterDependencies = obj => obj.dependencies ? obj.dependencies : {};

  /**
   * @description Gathers contract addresses by network
   * @param network - Network to gather addresses for
   * @param fallbackNetwork - Optional fallback network for fork networks
   * @returns Contract names by addresses mapping e.g { "HelloWorld": "0x123" }
   */
  const mapContractAliasesToNetworkAddress = (network, fallbackNetwork) => contracts => {
    return Object.entries(contracts).reduce((c, _ref) => {
      let [key, value] = _ref;
      const alias = value?.aliases?.[network] || fallbackNetwork && value?.aliases?.[fallbackNetwork];
      if (alias) {
        c[key] = alias;
      }
      return c;
    }, {});
  };

  /**
   * @description Gathers dependency addresses by network
   * @param network - Network to gather addresses for
   * @param fallbackNetwork - Optional fallback network for fork networks
   * @returns Dependency names by addresses mapping e.g { "HelloWorld": "0x123" }
   */
  const mapDependencyAliasesToNetworkAddress = (network, fallbackNetwork) => dependencies => {
    return Object.entries(dependencies).reduce((c, _ref2) => {
      let [key, value] = _ref2;
      const alias = value?.aliases?.[network] || fallbackNetwork && value?.aliases?.[fallbackNetwork];
      if (alias) {
        c[key] = alias;
      }
      return c;
    }, {});
  };

  /**
   * @description Gathers contract canonical references for import aliases
   * @param contracts - Contracts from flow.json
   * @returns Contract canonical names mapping e.g { "FUSD1.canonical": "FUSD" }
   *
   * Import aliases allow multiple deployments of the same contract to different addresses.
   * The canonical field specifies the actual contract name in the source code.
   *
   * Example:
   * - Input: { FUSD1: { canonical: "FUSD" } }
   * - Output: { "FUSD1.canonical": "FUSD" }
   * - Config stored as: system.contracts.FUSD1.canonical = "FUSD"
   * - Resolves to: import FUSD as FUSD1 from 0x...
   */
  const mapContractCanonicals = contracts => {
    return Object.entries(contracts).reduce((c, _ref3) => {
      let [key, value] = _ref3;
      if (value?.canonical) {
        c[`${key}.canonical`] = value.canonical;
      }
      return c;
    }, {});
  };

  /**
   * @description Gathers dependency canonical references for import aliases
   * @param dependencies - Dependencies from flow.json
   * @returns Dependency canonical names mapping e.g { "FungibleTokenV2.canonical": "FungibleToken" }
   *
   * Import aliases allow multiple deployments of the same contract to different addresses.
   * The canonical field specifies the actual contract name in the source code.
   *
   * Example:
   * - Input: { FungibleTokenV2: { canonical: "FungibleToken" } }
   * - Output: { "FungibleTokenV2.canonical": "FungibleToken" }
   * - Config stored as: system.contracts.FungibleTokenV2.canonical = "FungibleToken"
   * - Resolves to: import FungibleToken as FungibleTokenV2 from 0x...
   */
  const mapDependencyCanonicals = dependencies => {
    return Object.entries(dependencies).reduce((c, _ref4) => {
      let [key, value] = _ref4;
      if (value?.canonical) {
        c[`${key}.canonical`] = value.canonical;
      }
      return c;
    }, {});
  };
  const mapDeploymentsToNetworkAddress = network => _ref5 => {
    let {
      deployments = {},
      accounts = {}
    } = _ref5;
    const networkDeployment = deployments?.[network];
    if (!networkDeployment) return {};
    return Object.entries(networkDeployment).reduce((c, _ref6) => {
      let [key, value] = _ref6;
      // Resolve account address
      const accountAddress = accounts[key]?.address;
      if (!accountAddress) return c;

      // Create an object assigning the address to the contract name.
      return value.reduce((acc, contract) => {
        return {
          ...acc,
          [contract]: accountAddress
        };
      }, c);
    }, {});
  };

  /**
   * @description Take in flow.json files and return contract to address mapping by network
   * @param jsons - Flow JSON or array of Flow JSONs
   * @param network - Network to gather addresses for (can be a fork network)
   * @returns Contract names by addresses mapping including canonical references
   *
   * Returns both contract addresses and canonical references for import aliases:
   * - Contract addresses: { "HelloWorld": "0x123", "FUSD1": "0xe223..." }
   * - Canonical references: { "FUSD1.canonical": "FUSD" }
   *
   * These are stored in config as:
   * - system.contracts.FUSD1 = "0xe223..."
   * - system.contracts.FUSD1.canonical = "FUSD"
   *
   * The canonical reference enables import alias resolution:
   * - import "FUSD1" → import FUSD as FUSD1 from 0xe223...
   *
   * Fork networks use fallback resolution:
   * - First collects all aliases from the parent network (e.g., "mainnet")
   * - Then overlays aliases from the fork network (e.g., "mainnet-fork")
   * - This allows fork-specific overrides while inheriting the rest
   */
  const getContracts = (jsons, network) => {
    const mergedJson = mergeFlowJSONs(jsons);

    // Determine fallback network for fork networks
    const networkConfig = mergedJson.networks?.[network];
    const fallbackNetwork = networkConfig && typeof networkConfig === "object" && "fork" in networkConfig ? networkConfig.fork : undefined;
    return mergePipe(mapDeploymentsToNetworkAddress(network), pipe(filterContracts, mapContractAliasesToNetworkAddress(network, fallbackNetwork)), pipe(filterDependencies, mapDependencyAliasesToNetworkAddress(network, fallbackNetwork)), pipe(filterContracts, mapContractCanonicals), pipe(filterDependencies, mapDependencyCanonicals))(mergedJson);
  };

  /**
   * @description Checks if string is hexidecimal
   * @param str - String to check
   * @returns Is hexidecimal status
   */
  const isHexidecimal = str => {
    // Check that it is a string
    if (typeof str !== "string") return false;
    return /^[0-9A-Fa-f]+$/.test(str);
  };

  /**
   * @description Checks flow.json file for private keys
   * @param flowJSON - Flow JSON
   * @returns Has private keys status
   */
  const hasPrivateKeys = flowJSON => {
    return Object.entries(flowJSON?.accounts ?? []).reduce((hasPrivateKey, _ref7) => {
      let [, value] = _ref7;
      if (hasPrivateKey) return true;
      return value && Object.prototype.hasOwnProperty.call(value, "key") && isHexidecimal(value?.key);
    }, false);
  };

  /**
   * @description Take in flow.json or array of flow.json files and checks for private keys
   * @param value - Flow JSON or array of Flow JSONs
   * @returns Has private keys status
   */
  const anyHasPrivateKeys = value => {
    if (Array.isArray(value)) return value.some(hasPrivateKeys);
    return hasPrivateKeys(value);
  };

  /**
   * @description Format network name, converting 'local' to 'emulator'
   * @param network - Network to format
   * @returns Formatted network name
   *
   * This function now accepts any network name (including custom fork networks).
   * It only transforms "local" to "emulator" for backward compatibility.
   */
  const cleanNetwork = network => {
    return network?.toLowerCase() === "local" ? "emulator" : network?.toLowerCase();
  };

  // Inject config into logger to break circular dependency
  logger__namespace.setConfig(config);
  const NAME = "config";
  const PUT = "PUT_CONFIG";
  const GET = "GET_CONFIG";
  const GET_ALL = "GET_ALL_CONFIG";
  const UPDATE = "UPDATE_CONFIG";
  const DELETE = "DELETE_CONFIG";
  const CLEAR = "CLEAR_CONFIG";
  const WHERE = "WHERE_CONFIG";
  const UPDATED = "CONFIG/UPDATED";
  const identity = v => v;
  const HANDLERS = {
    [PUT]: (ctx, _letter, _ref) => {
      let {
        key,
        value
      } = _ref;
      if (key == null) throw new Error("Missing 'key' for config/put.");
      ctx.put(key, value);
      ctx.broadcast(UPDATED, {
        ...ctx.all()
      });
    },
    [GET]: (ctx, letter, _ref2) => {
      let {
        key,
        fallback
      } = _ref2;
      if (key == null) throw new Error("Missing 'key' for config/get");
      letter.reply(ctx.get(key, fallback));
    },
    [GET_ALL]: (ctx, letter) => {
      letter.reply({
        ...ctx.all()
      });
    },
    [UPDATE]: (ctx, letter, _ref3) => {
      let {
        key,
        fn
      } = _ref3;
      if (key == null) throw new Error("Missing 'key' for config/update");
      ctx.update(key, fn || identity);
      ctx.broadcast(UPDATED, {
        ...ctx.all()
      });
    },
    [DELETE]: (ctx, letter, _ref4) => {
      let {
        key
      } = _ref4;
      if (key == null) throw new Error("Missing 'key' for config/delete");
      ctx.delete(key);
      ctx.broadcast(UPDATED, {
        ...ctx.all()
      });
    },
    [CLEAR]: ctx => {
      const keys = Object.keys(ctx.all());
      for (const key of keys) ctx.delete(key);
      ctx.broadcast(UPDATED, {
        ...ctx.all()
      });
    },
    [WHERE]: (ctx, letter, _ref5) => {
      let {
        pattern
      } = _ref5;
      if (pattern == null) throw new Error("Missing 'pattern' for config/where");
      letter.reply(ctx.where(pattern));
    },
    [utilActor.SUBSCRIBE]: (ctx, letter) => {
      ctx.subscribe(letter.from);
      ctx.send(letter.from, UPDATED, {
        ...ctx.all()
      });
    },
    [utilActor.UNSUBSCRIBE]: (ctx, letter) => {
      ctx.unsubscribe(letter.from);
    }
  };
  utilActor.spawn(HANDLERS, NAME);

  /**
   * @description Adds a key-value pair to the config
   * @param key - The key to add
   * @param value - The value to add
   * @returns The config object
   */
  function put(key, value) {
    utilActor.send(NAME, PUT, {
      key,
      value
    });
    return config();
  }

  /**
   * @description Gets a key-value pair with a fallback from the config
   * @param key - The key to add
   * @param fallback - The fallback value to return if key is not found
   * @returns The value found at key or fallback
   */
  function get(key, fallback) {
    return utilActor.send(NAME, GET, {
      key,
      fallback
    }, {
      expectReply: true,
      timeout: 10
    });
  }

  /**
   * @description Returns the first non null config value or the fallback
   * @param wants - The keys to search for
   * @param fallback - The fallback value to return if key is not found
   * @returns The value found at key or fallback
   */
  async function first() {
    let wants = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
    let fallback = arguments.length > 1 ? arguments[1] : undefined;
    if (!wants.length) return fallback;
    const [head, ...rest] = wants;
    const ret = await get(head);
    if (ret == null) return first(rest, fallback);
    return ret;
  }

  /**
   * @description Returns the current config
   * @returns The config object
   */
  function all() {
    return utilActor.send(NAME, GET_ALL, null, {
      expectReply: true,
      timeout: 10
    });
  }

  /**
   * @description Updates a key-value pair in the config
   * @param key - The key to update
   * @param fn - The function to update the value with
   * @returns The config object
   */
  function update(key) {
    let fn = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : identity;
    utilActor.send(NAME, UPDATE, {
      key,
      fn
    });
    return config();
  }

  /**
   * @description Deletes a key-value pair from the config
   * @param key - The key to delete
   * @returns The config object
   */
  function _delete(key) {
    utilActor.send(NAME, DELETE, {
      key
    });
    return config();
  }

  /**
   * @description Returns a subset of the config based on a pattern
   * @param pattern - The pattern to match keys against
   * @returns The subset of the config
   */
  function where(pattern) {
    return utilActor.send(NAME, WHERE, {
      pattern
    }, {
      expectReply: true,
      timeout: 10
    });
  }

  /**
   * @description Subscribes to config updates
   * @param callback - The callback to call when config is updated
   * @returns The unsubscribe function
   */
  function subscribe(callback) {
    return utilActor.subscriber(NAME, () => utilActor.spawn(HANDLERS, NAME), callback);
  }

  /**
   * @description Clears the config
   */
  async function clearConfig() {
    await utilActor.send(NAME, CLEAR);
  }

  /**
   * @description Resets the config to a previous state
   * @param oldConfig - The previous config state
   * @returns The config object
   */
  async function resetConfig(oldConfig) {
    return clearConfig().then(() => config(oldConfig));
  }

  /**
   * @description Takes in flow.json or array of flow.json files and creates contract placeholders
   * @param data - The data to load
   * @param data.flowJSON - The flow.json or array of flow.json files
   * @param options - override flag
   * @param options.ignoreConflicts - ignore conflicts and override config
   */
  async function load(data) {
    let {
      ignoreConflicts = false
    } = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    const network = await get("flow.network");
    const cleanedNetwork = cleanNetwork(network);
    const {
      flowJSON
    } = data;
    utilInvariant.invariant(Boolean(flowJSON), "config.load -- 'flowJSON' must be defined");
    utilInvariant.invariant(!!cleanedNetwork, `Flow Network Required -- In order for FCL to load your contracts please define "flow.network" to "emulator", "local", "testnet", or "mainnet" in your config. See more here: https://developers.flow.com/tools/fcl-js/reference/configure-fcl`);
    if (anyHasPrivateKeys(flowJSON)) {
      const isEmulator = cleanedNetwork === "emulator";
      logger__namespace.log({
        title: "Private Keys Detected",
        message: `Private keys should be stored in a separate flow.json file for security. See more here: https://developers.flow.com/tools/flow-cli/security`,
        level: isEmulator ? logger__namespace.LEVELS.warn : logger__namespace.LEVELS.error
      });
      utilInvariant.invariant(isEmulator, `Private keys should be stored in a separate flow.json file for security. See more here: https://developers.flow.com/tools/flow-cli/security`);
    }
    for (const [key, value] of Object.entries(getContracts(flowJSON, cleanedNetwork))) {
      const contractConfigKey = `0x${key}`;
      const existingContractConfigKey = await get(contractConfigKey);
      if (existingContractConfigKey && existingContractConfigKey !== value && !ignoreConflicts) {
        logger__namespace.log({
          title: "Contract Placeholder Conflict Detected",
          message: `A generated contract placeholder from config.load conflicts with a placeholder you've set manually in config have the same name.`,
          level: logger__namespace.LEVELS.warn
        });
      } else {
        put(contractConfigKey, value);
      }
      const systemContractConfigKey = `system.contracts.${key}`;
      const systemExistingContractConfigKeyValue = await get(systemContractConfigKey);
      if (systemExistingContractConfigKeyValue && systemExistingContractConfigKeyValue !== value && !ignoreConflicts) {
        logger__namespace.log({
          title: "Contract Placeholder Conflict Detected",
          message: `A generated contract placeholder from config.load conflicts with a placeholder you've set manually in config have the same name.`,
          level: logger__namespace.LEVELS.warn
        });
      } else {
        put(systemContractConfigKey, value);
      }
    }
  }

  /**
   * @description Sets the config
   * @param values - The values to set
   * @returns The config object
   */
  function config(values) {
    if (values != null && typeof values === "object") {
      Object.keys(values).map(d => put(d, values[d]));
    }
    return {
      put,
      get,
      all,
      first,
      update,
      delete: _delete,
      where,
      subscribe,
      overload,
      load
    };
  }
  config.put = put;
  config.get = get;
  config.all = all;
  config.first = first;
  config.update = update;
  config.delete = _delete;
  config.where = where;
  config.subscribe = subscribe;
  config.overload = overload;
  config.load = load;

  /**
   * @description Temporarily overloads the config with the given values and calls the callback
   * @param values - The values to overload the config with
   * @param callback - The callback to call with the overloaded config
   * @returns The result of the callback
   */
  async function overload(values, callback) {
    const oldConfig = await all();
    try {
      config(values);
      const result = await callback(await all());
      return result;
    } finally {
      await resetConfig(oldConfig);
    }
  }

  exports.clearConfig = clearConfig;
  exports.config = config;
  exports.getContracts = getContracts;

}));
